Avoid unnecessary UTF-8 string generation

Constellation opened the pull request:
#### 0e2f9b02c86cd64041171e370c18f4b262ea7505
<pre>
Avoid unnecessary UTF-8 string generation
<a href="https://bugs.webkit.org/show_bug.cgi?id=243931">https://bugs.webkit.org/show_bug.cgi?id=243931</a>

Reviewed by Darin Adler.

This patch removes unnecessary UTF-8 string generation by using tryGetUtf8ForRange.
We consolidate the implementation into StringImpl::tryGetUtf8ForCharacters and
implement the other functions by using that.

One of the important change is that we avoid unnecessary allocation in ScriptBuffer.
This is important since ScriptBuffer size tends to be very large.

* Source/JavaScriptCore/jsc.cpp:
(toCString):
(dumpException):
(runInteractive):
* Source/JavaScriptCore/runtime/JSDateMath.cpp:
(JSC::DateCache::parseDate):
* Source/WTF/wtf/PrintStream.cpp:
(WTF::printInternal):
* Source/WTF/wtf/URL.cpp:
(WTF::percentEncodeCharacters):
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::utf8ForCharacters):
(WTF::StringImpl::tryGetUTF8ForRange const):
(WTF::StringImpl::tryGetUTF8 const):
(WTF::StringImpl::utf8 const):
(WTF::StringImpl::tryGetUtf8ForRange const): Deleted.
(WTF::StringImpl::tryGetUtf8 const): Deleted.
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::tryGetUTF8ForRange const):
(WTF::StringImpl::tryGetUTF8ForCharacters):
(WTF::StringImpl::tryGetUtf8ForRange const): Deleted.
* Source/WTF/wtf/text/StringView.cpp:
(WTF::StringView::tryGetUTF8 const):
(WTF::StringView::utf8 const):
(WTF::StringView::tryGetUtf8 const): Deleted.
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::tryGetUTF8ForRange const):
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::String::tryGetUTF8 const):
(WTF::String::utf8 const):
(WTF::String::tryGetUtf8 const): Deleted.
* Source/WTF/wtf/text/WTFString.h:
(WTF::String::tryGetUTF8ForRange const):
* Source/WebCore/Modules/webauthn/WebAuthenticationUtils.cpp:
(WebCore::produceRpIdHash):
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::processKeyframeLikeObject):
* Source/WebCore/animation/WebAnimationUtilities.cpp:
(WebCore::compareCSSTransitions):
* Source/WebCore/crypto/SubtleCrypto.cpp:
(WebCore::SubtleCrypto::wrapKey):
* Source/WebCore/dom/TextDecoder.cpp:
(WebCore::TextDecoder::TextDecoder):
(WebCore::TextDecoder::create):
* Source/WebCore/dom/TextDecoder.h:
(WebCore::TextDecoder::fatal const): Deleted.
(WebCore::TextDecoder::ignoreBOM const): Deleted.
* Source/WebCore/dom/TextEncoder.cpp:
(WebCore::TextEncoder::encode const):
* Source/WebCore/platform/generic/KeyedEncoderGeneric.cpp:
(WebCore::KeyedEncoderGeneric::encodeString):
* Source/WebCore/platform/win/ClipboardUtilitiesWin.cpp:
(WebCore::getUTF8Data):
(WebCore::setUTF8Data):
(WebCore::getClipboardMap):
(WebCore::getUtf8Data): Deleted.
(WebCore::setUtf8Data): Deleted.
* Source/WebCore/workers/ScriptBuffer.cpp:
(WebCore::ScriptBuffer::ScriptBuffer):
(WebCore::ScriptBuffer::append):
* Source/WebCore/workers/service/server/SWScriptStorage.cpp:
(WebCore::SWScriptStorage::sha2Hash const):
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::XMLDocumentParser::appendFragmentSource):
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp:
(WebKit::CtapAuthenticator::continueGetPinTokenAfterRequestPin):
* Tools/DumpRenderTree/mac/DumpRenderTree.mm:
(dumpFramesAsText):
* Tools/DumpRenderTree/win/UIDelegate.cpp:
(toMessage):
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp:
(WTR::InjectedBundle::dumpToStdErr):
(WTR::InjectedBundle::outputText):

Canonical link: <a href="https://commits.webkit.org/253642@main">https://commits.webkit.org/253642@main</a>
</pre>

<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/5125abc6325ae90210b6ff97c1761d8ccf3dda46

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/86648 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/30701 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/43/builds/17585 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/8/builds/95491 "Built successfully") | [✅ 🛠 🧪 win](https://ews-build.webkit.org/#/builders/10/builds/149220 "Built successfully and passed tests") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/11/builds/90633 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/29078 "Built successfully") | [✅ 🛠 mac-debug](https://ews-build.webkit.org/#/builders/71/builds/25514 "Built successfully") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/78829 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/12/builds/90734 "Built successfully") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/92264 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/68/builds/23499 "Passed tests") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/73572 "Built successfully") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/35/builds/23566 "Passed tests") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/9/builds/78504 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/3/builds/78855 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/34/builds/66572 "Passed tests") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/1/builds/78583 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/26864 "Built successfully") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/73/builds/12707 "Passed tests") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/38/builds/72220 "Built successfully") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/26783 "Built successfully") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/70/builds/13722 "Passed tests") | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/46/builds/25789 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/74/builds/2584 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/28461 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/60/builds/36581 "Passed tests") | [✅ 🛠 jsc-mips](https://ews-build.webkit.org/#/builders/37/builds/75003 "Built successfully") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/28405 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/62/builds/33001 "Passed tests") | [✅ 🧪 jsc-mips-tests](https://ews-build.webkit.org/#/builders/45/builds/16589 "Passed tests") | 
<!--EWS-Status-Bubble-End-->

darinadler commented on Source/WTF/wtf/URL.cpp:
```diff
@@ -531,20 +531,21 @@ template<typename StringType>
+            }
+            return builder.toString();
+        }, 0, input.length());
+        RELEASE_ASSERT(result);
```
Why did we need this to be a release assertion? Is there something dangerous about a null `String`?

Constellation replied:
This code is aligning to the old `utf8()` call here.

darinadler commented on Source/WTF/wtf/text/StringImpl.h:
```diff
@@ -323,6 +323,11 @@ class StringImpl : private StringImplShape {
     static WTF_EXPORT_PRIVATE Expected<CString, UTF8ConversionError> utf8ForCharacters(const UChar* characters, unsigned length, ConversionMode = LenientConversion);
 
+    template<typename Func>
+    static Expected<std::invoke_result_t<Func, Span<const char>>, UTF8ConversionError> tryGetUtf8ForCharacters(const Func&, const LChar* characters, unsigned length);
```
This should be `tryGetUTF8…` I don’t know why other identifiers are capitalizing it `Utf`. It is easier to spot when it has the traditional all capital style.

Constellation replied:
Updated.

darinadler commented on Source/WTF/wtf/text/StringImpl.h:
```diff
@@ -1364,6 +1369,14 @@ inline Expected<std::invoke_result_t<Func, Span<const char>>, UTF8ConversionErro
     ASSERT(offset + length <= this->length());
 
+    if (is8Bit())
+        return StringImpl::tryGetUtf8ForCharacters(function, characters8() + offset, length);
```
No need for `StringImpl` prefix here?

Constellation replied:
Dropped.

darinadler commented on Source/WTF/wtf/text/StringImpl.h:
```diff
@@ -1382,35 +1395,44 @@ inline Expected<std::invoke_result_t<Func, Span<const char>>, UTF8ConversionErro
+inline Expected<std::invoke_result_t<Func, Span<const char>>, UTF8ConversionError> StringImpl::tryGetUtf8ForCharacters(const Func& function, const UChar* characters, unsigned length, ConversionMode mode)
+{
+    if (!length) {
+        const char* const emptyString = "";
```
Use `constexpr` here?

Constellation replied:
Added.

darinadler commented on Source/WTF/wtf/text/StringView.h:
```diff
@@ -124,6 +124,9 @@ class StringView final {
     WTF_EXPORT_PRIVATE CString utf8(ConversionMode = LenientConversion) const;
 
+    template<typename Func>
+    Expected<std::invoke_result_t<Func, Span<const char>>, UTF8ConversionError> tryGetUtf8ForRange(const Func&, unsigned offset, unsigned length, ConversionMode = LenientConversion) const;
```
Same thought on `Utf` here.

Constellation replied:
Changed.

darinadler commented on Source/WTF/wtf/text/StringImpl.h:
```diff
@@ -1364,6 +1369,14 @@ inline Expected<std::invoke_result_t<Func, Span<const char>>, UTF8ConversionErro
 
+    if (is8Bit())
+        return StringImpl::tryGetUtf8ForCharacters(function, characters8() + offset, length);
+    return StringImpl::tryGetUtf8ForCharacters(function, characters16() + offset, length, mode);
```
Same here.

darinadler commented on Source/WTF/wtf/text/WTFString.h:
```diff
@@ -132,6 +132,8 @@ class String final {
     WTF_EXPORT_PRIVATE CString utf8() const;
 
+    template<typename Func>
+    Expected<std::invoke_result_t<Func, Span<const char>>, UTF8ConversionError> tryGetUtf8ForRange(const Func&, unsigned offset, unsigned length, ConversionMode = LenientConversion) const;
```
More `Utf`.

darinadler commented on Source/WTF/wtf/text/WTFString.h:
```diff
@@ -512,6 +514,18 @@ inline void String::clearImplIfNotShared()
+    ASSERT(offset <= this->length());
+    ASSERT(offset + length <= this->length());
+    if (!m_impl) {
+        const char* const emptyString = "";
```
`constexpr`

Constellation replied:
Added.

darinadler commented on Source/WTF/wtf/text/WTFString.h:
```diff
@@ -512,6 +514,18 @@ inline void String::clearImplIfNotShared()
+inline Expected<std::invoke_result_t<Func, Span<const char>>, UTF8ConversionError> String::tryGetUtf8ForRange(const Func& function, unsigned offset, unsigned length, ConversionMode mode) const
+{
+    ASSERT(offset <= this->length());
+    ASSERT(offset + length <= this->length());
```
This way of writing the assertion is vulnerable to overflow if `length` is very large. Not critical but better to write it in a way that is not. One way is `ASSERT(length <= this->length() - offset)` but I am sure there are many others.

Constellation replied:
Fixed.

darinadler commented on Source/WTF/wtf/text/WTFString.h:
```diff
@@ -512,6 +514,18 @@ inline void String::clearImplIfNotShared()
+    ASSERT(offset + length <= this->length());
+    if (!m_impl) {
+        const char* const emptyString = "";
+        return function(Span<const char> { emptyString, emptyString });
```
Can we write just `Span` here or does it not deduce the type?

Constellation replied:
Changed.

darinadler commented on Source/WTF/wtf/text/StringImpl.h:
```diff
@@ -1382,35 +1395,44 @@ inline Expected<std::invoke_result_t<Func, Span<const char>>, UTF8ConversionErro
+{
+    if (!length) {
+        const char* const emptyString = "";
+        return function(Span<const char> { emptyString, emptyString });
```
Can we write just `Span` here or does it not deduce the type?

Constellation replied:
Changed.

darinadler commented on Source/WTF/wtf/text/StringView.h:
```diff
@@ -1418,6 +1421,16 @@ inline bool AtomString::endsWithIgnoringASCIICase(StringView string) const
 }
 
+template<typename Func>
+inline Expected<std::invoke_result_t<Func, Span<const char>>, UTF8ConversionError> StringView::tryGetUtf8ForRange(const Func& function, unsigned offset, unsigned length, ConversionMode mode) const
```
`Utf`

Constellation replied:
Changed.

darinadler reviewed (approved):
Looks great. Would like to measure performance impact somehow.

Constellation committed:
Avoid unnecessary UTF-8 string generation
https://bugs.webkit.org/show_bug.cgi?id=243931

Reviewed by Darin Adler.

This patch removes unnecessary UTF-8 string generation by using tryGetUtf8ForRange.
We consolidate the implementation into StringImpl::tryGetUtf8ForCharacters and
implement the other functions by using that.

One of the important change is that we avoid unnecessary allocation in ScriptBuffer.
This is important since ScriptBuffer size tends to be very large.

* Source/JavaScriptCore/jsc.cpp:
(toCString):
(dumpException):
(runInteractive):
* Source/JavaScriptCore/runtime/JSDateMath.cpp:
(JSC::DateCache::parseDate):
* Source/WTF/wtf/PrintStream.cpp:
(WTF::printInternal):
* Source/WTF/wtf/URL.cpp:
(WTF::percentEncodeCharacters):
* Source/WTF/wtf/text/StringImpl.cpp:
(WTF::StringImpl::utf8ForCharacters):
(WTF::StringImpl::tryGetUTF8ForRange const):
(WTF::StringImpl::tryGetUTF8 const):
(WTF::StringImpl::utf8 const):
(WTF::StringImpl::tryGetUtf8ForRange const): Deleted.
(WTF::StringImpl::tryGetUtf8 const): Deleted.
* Source/WTF/wtf/text/StringImpl.h:
(WTF::StringImpl::tryGetUTF8ForRange const):
(WTF::StringImpl::tryGetUTF8ForCharacters):
(WTF::StringImpl::tryGetUtf8ForRange const): Deleted.
* Source/WTF/wtf/text/StringView.cpp:
(WTF::StringView::tryGetUTF8 const):
(WTF::StringView::utf8 const):
(WTF::StringView::tryGetUtf8 const): Deleted.
* Source/WTF/wtf/text/StringView.h:
(WTF::StringView::tryGetUTF8ForRange const):
* Source/WTF/wtf/text/WTFString.cpp:
(WTF::String::tryGetUTF8 const):
(WTF::String::utf8 const):
(WTF::String::tryGetUtf8 const): Deleted.
* Source/WTF/wtf/text/WTFString.h:
(WTF::String::tryGetUTF8ForRange const):
* Source/WebCore/Modules/webauthn/WebAuthenticationUtils.cpp:
(WebCore::produceRpIdHash):
* Source/WebCore/animation/KeyframeEffect.cpp:
(WebCore::processKeyframeLikeObject):
* Source/WebCore/animation/WebAnimationUtilities.cpp:
(WebCore::compareCSSTransitions):
* Source/WebCore/crypto/SubtleCrypto.cpp:
(WebCore::SubtleCrypto::wrapKey):
* Source/WebCore/dom/TextDecoder.cpp:
(WebCore::TextDecoder::TextDecoder):
(WebCore::TextDecoder::create):
* Source/WebCore/dom/TextDecoder.h:
(WebCore::TextDecoder::fatal const): Deleted.
(WebCore::TextDecoder::ignoreBOM const): Deleted.
* Source/WebCore/dom/TextEncoder.cpp:
(WebCore::TextEncoder::encode const):
* Source/WebCore/platform/generic/KeyedEncoderGeneric.cpp:
(WebCore::KeyedEncoderGeneric::encodeString):
* Source/WebCore/platform/win/ClipboardUtilitiesWin.cpp:
(WebCore::getUTF8Data):
(WebCore::setUTF8Data):
(WebCore::getClipboardMap):
(WebCore::getUtf8Data): Deleted.
(WebCore::setUtf8Data): Deleted.
* Source/WebCore/workers/ScriptBuffer.cpp:
(WebCore::ScriptBuffer::ScriptBuffer):
(WebCore::ScriptBuffer::append):
* Source/WebCore/workers/service/server/SWScriptStorage.cpp:
(WebCore::SWScriptStorage::sha2Hash const):
* Source/WebCore/xml/parser/XMLDocumentParserLibxml2.cpp:
(WebCore::XMLDocumentParser::appendFragmentSource):
* Source/WebKit/UIProcess/WebAuthentication/fido/CtapAuthenticator.cpp:
(WebKit::CtapAuthenticator::continueGetPinTokenAfterRequestPin):
* Tools/DumpRenderTree/mac/DumpRenderTree.mm:
(dumpFramesAsText):
* Tools/DumpRenderTree/win/UIDelegate.cpp:
(toMessage):
* Tools/WebKitTestRunner/InjectedBundle/InjectedBundle.cpp:
(WTR::InjectedBundle::dumpToStdErr):
(WTR::InjectedBundle::outputText):

Canonical link: https://commits.webkit.org/253642@main
